crates/biome_js_analyze/src/assists/source/organize_imports/util.rs

```diff
@@ -0,0 +1,380 @@
+                PathComponent::RootDir,
+            ) => Ordering::Greater,
+            (PathComponent::CurDir, PathComponent::Normal(_)) => Ordering::Less,
+            (PathComponent::Normal(_), PathComponent::CurDir) => Ordering::Greater,
+            (PathComponent::Normal(s1), PathComponent::Normal(s2)) => {
+                ImportSourceAsciiCollator.cmp_osstr(s1, s2)
+            }
+        }
+    }
+}
+
+/// This type is analog to [std::path::Components] with the following changes:
+/// - The iterator yields [PathComponent] instead of [std::path::Component].
+/// - Consecutive parent directories such as `../..` yields a single [PathComponent::ParentDir] with
+///   the count of parent directories.
+/// - `./..` is normalized to `..`
+struct PathComponents<'a> {
+    inner: std::path::Components<'a>,
+}
+impl<'a> From<&'a Path> for PathComponents<'a> {
+    fn from(path: &'a Path) -> Self {
+        Self {
+            inner: path.components(),
+        }
+    }
+}
+impl<'a> Iterator for PathComponents<'a> {
+    type Item = PathComponent<'a>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        Some(match self.inner.next()? {
+            Component::Prefix(c) => PathComponent::Prefix(c),
+            Component::RootDir => PathComponent::RootDir,
+            Component::ParentDir => {
+                let count = self
+                    .inner
+                    .clone()
+                    .take_while(|c| matches!(c, Component::ParentDir))
+                    .count();
+                for _ in 1..=count {
+                    self.inner.next();
+                }
+                PathComponent::ParentDir(count + 1)
+            }
+            Component::CurDir => {
+                // Normalize `./../..` to `../..`.
+                // Note that [std::path::Components] already normalizes `.././..` to `../..`
+                let parent_dir_count = self
+                    .inner
+                    .clone()
+                    .take_while(|c| matches!(c, Component::ParentDir))
+                    .count();
+                if parent_dir_count == 0 {
+                    PathComponent::CurDir
+                } else {
+                    for _ in 1..=parent_dir_count {
+                        self.inner.next();
+                    }
+                    PathComponent::ParentDir(parent_dir_count)
+                }
+            }
+            Component::Normal(s) => PathComponent::Normal(s),
+        })
+    }
+}
+
+/// Custom collation order to get a natural irder between import sources.
+///
+/// Non-printable characters and alphanumeric characters have the same order as [biome_string_case::CldrAsciiCollator].
+/// Others are ordered differently, but still between non-printable characters and alphanumeric characters.
+pub struct ImportSourceAsciiCollator;
+impl ImportSourceAsciiCollator {
+    const COLLATION: [u8; 128] = [
```

Comment:
Isn't this table the same we have in the other file?

Reply:
Another approach is to compare two import sources with the same kind as paths and to use the natural order defined in `biome_string_case`. It is less natural than the one I created in here but should be good for most of the cases.

Reply:
The [Unicode Collation Algorithm](https://www.unicode.org/reports/tr10/) allows ordering ain a more natural way characters.
The algorithm is a general and must respect some conformance rules.
It provides a default collation table named DUCET (Default Unicode Collation Element Table) that can be used as a base for any collation. It is expected that a locale defines a custom collation order for some characters and uses DUCET for the ones it doesn't care.

It seems that DUCET has been supplanted by a new collation table named [CLDR](https://en.wikipedia.org/wiki/Common_Locale_Data_Repository) (Common Locale Data Repository). DUCET and CLDR are mostly identical. In the ASCII side, only the order between punctuation and symbols is different between DUCET and CLDR. Nowadays, most of modern implementation relies on CLDR as a basic collation.

`CldrAsciiCollator` is CLDR reduced to the ASCII characters and extended to any byte by mapping all bytes higher than `0x7f` to themselves.

`ImportSourceAsciiCollator` is basically `CldrAsciiCollator` but with a different order for punctuation and symbols.
Let's take an example to highlight the differences between `CldrAsciiCollator` and `ImportSourceAsciiCollator`.
Given the following list of strings (sorted according to their binary representation):

```
/a-b/
/a/b/
https://example.org/path?prop-1=val
https://example.org/path?prop=val
npm:@scope/package@^10.0/path
npm:@scope/package@>=10.0/path
npm:@scope/package@10.0/path
npm:@scope/package@9.0/path
npm:@scope/package/path
```

Using `CldrAsciiCollator`, we obtain the following sorted list:

```
/a-b/
/a/b/
https://example.org/path?prop-1=val
https://example.org/path?prop=val
npm:@scope/package@^10.0/path
npm:@scope/package@>=10.0/path
npm:@scope/package@9.0/path
npm:@scope/package@10.0/path
npm:@scope/package/path
```

Using `ImportSourceAsciiCollator`, we obtain the following sorted list:

```
/a/b/
/a-b/
https://example.org/path?prop=val
https://example.org/path?prop-1=val
npm:@scope/package/path
npm:@scope/package@>=10.0/path
npm:@scope/package@^10.0/path
npm:@scope/package@9.0/path
npm:@scope/package@10.0/path
```

`ImportSourceAsciiCollator` is designed to provide a natural ordering of paths and URLs.
As you can note, `/a/b` is placed before `/a-b/` because `/` < `-` (while `-` < `/` in CLDR).
We have ordering of semantic specifier (`>` < `^` < `~`).
We are URL-friendly (`?` < `#` < `=` < `&`) that allows a good order between URLs with query and/or fragment parts.

---

I am thinking about comparing all imports of a same kind as paths. i.e. we first decompose an import into path components. This will ensure that `/` has the priority. If we make this change, we could reduce our dependency on `ImportSourceAsciiCollator` or even drop it and directly use `CldrAsciiCollator`.

Reply:
> Note: for now no code is using CldrAsciiCollator. It is here for potential future uses and to provide an example to how to implement AsciiCollator.

Thank you for the detailed explanation. I don't see any particular value to add code that won't be used (the `CldrAsciiCollator` won't be). I can't see any future usage as well.

Reply:
> I can't see any future usage as well.

Actually I see a usage where we actually use natord (for example `useSortedKeys`).

Reply:
Awesome, thank you! Happy to merge the PR